### DISCUSSION TITLE
Switch for loops in ApproxBetweeness for score aggregation?

### DISCUSSION THREAD

**Discussion by rionda** [networkit/networkit#684] [Ideas]
https://github.com/networkit/networkit/blob/ecb1af8c2c6ef3c484b45cdb9fdead584cbd94bc/networkit/cpp/centrality/ApproxBetweenness.cpp#L108

 https://github.com/networkit/networkit/blob/ecb1af8c2c6ef3c484b45cdb9fdead584cbd94bc/networkit/cpp/centrality/ApproxBetweenness.cpp#L109

(I admit this suggestion is not based on timing measurements).
 
There is a cost associated to the setup of an omp parallel loop. Rather than paying this cost as many times as there are processors/threads (i.e., local scores to be aggregated), switching the loops should allow to only pay the cost once.

(one also has to wonder whether there should be a special case that skips the for loops altogether if there is a single thread, and just assigns `local` to `scoreData`).

(thinking of the bigger picture, I suspect that this implementation can be reorganized to use an OMP reduction, avoiding this separate aggregation step, or better, incorporating it in the reduction)

Matteo

**Comment by angriman** [comment]
I think that switching the two loops makes sense. About the reduction, I am not sure how that would work since `scoreData` is a vector, so all threads need to work on their separate copy anyway. What about replacing
https://github.com/networkit/networkit/blob/ecb1af8c2c6ef3c484b45cdb9fdead584cbd94bc/networkit/cpp/centrality/ApproxBetweenness.cpp#L97
with an atomic operation? This would allow us to remove the `scorePerThread` vector and skip the aggregation step.

**Comment by rionda** [comment]
I'm not a fan: the atomic puts a lock at the vector level, rather than at the single element level. It's too coarse. I find reductions much more elegant, but YMMV. =)

**Comment by angriman** [comment]
Do you mean an OpenMP reduction? Because I am not sure how it could work here. Btw I compared the running time with 24 threads of swapping loops and atomic operations (here is the plot [atomic-agg.pdf](https://github.com/networkit/networkit/files/5222840/atomic-agg.pdf)) and also atomic operations vs the original implementation (here is the plot [atomic-orig.pdf](https://github.com/networkit/networkit/files/5222846/atomic-orig.pdf)).
Seems like, on overage, using atomic operations is the fastest solution -- probably in this case using additional thread-local memory is more expensive than locks.

**Comment by rionda** [comment]
So atomic-agg.pdf is atomic vs swapped loops, while atomic-orig.pdf is atomic vs the original implementation, right? Interesting. So just swapping loops speed things up to be "almost" there, but atomic is still a little faster. Indeed the reduction would likely be similar to the swapped loops.

I almost wonder whether it would be worth having a `std::vector`` of `omp_lock_t` variables, one per node, and only lock the one specific to a node before incrementing the score for that node, to have more fine-grained locking, thus reducing contention.

**Comment by michitux** [comment]
Are you sure this atomic update is implemented with a lock and not a CAS loop?  If this is indeed the case, wouldn't it be better to just implement the CAS loop? I imagine that rarely two threads access the same memory location and thus the CAS should almost never fail.

**Comment by rionda** [comment]
No, I don't know how different openmp implementation may implement atomic, and you are likely right. =)

**Comment by hmeyerhenke** [comment]
Let's take another look together when Alexander is back.

**Comment by angriman** [comment]
I think that the atomic update performs a `fetch-and-add` operation. I also implemented a cas loop using `compare_exchange_weak` to update the scores and the performance are very similar to atomic updates. I think we should go with atomic updates, among the three it's the simplest and the fastest solution.

**Comment by avdgrinten** [comment]
`compare_exchange_weak` doesn't make sense. Just use `fetch_add` (which is specifically made for this purpose).

**Comment by angriman** [comment]
I did not do that because I could not find a way to use `fetch_add` with `double`.

**Comment by michitux** [comment]
> I think that the atomic update performs a `fetch-and-add` operation. I also implemented a cas loop using `compare_exchange_weak` to update the scores and the performance are very similar to atomic updates. I think we should go with atomic updates, among the three it's the simplest and the fastest solution.

There is no such thing as a fetch-and-add operation for floating point numbers on x86-64 CPUs, see, e.g., [this question on Stack Overflow](https://stackoverflow.com/questions/45055402/atomic-double-floating-point-or-sse-avx-vector-load-store-on-x86-64), therefore the atomic update definitely does not use them. However, at least GCC and clang indeed use a compare-and-swap loop as [this example in the Compiler Explorer](https://godbolt.org/z/M3GP99) shows. Therefore, I think we should simply use `#pragma omp atomic update`.

> `compare_exchange_weak` doesn't make sense. Just use `fetch_add` (which is specifically made for this purpose).

As mentioned above, `fetch_add` is not available for floating point numbers on x86-64 CPUs.

**Comment by avdgrinten** [comment]
True, I was not aware of the fact that we're dealing with doubles here when I wrote my comment.

(Nitpick: `fetch_add` is a C++ standard function, its availability does not depend on the CPU. If the CPU doesn't implement it (e.g., on older ARMs), it will be emulated using CAS or locks. But you're right that pre-C++20, it is not available for doubles. This has nothing to do with the fact that x86 does not have it as an atomic instruction but rather with the issue that C floating point types are traditionally allowed to have padding bits that do not contribute to the `double`s value.)

By the way, the second answer in your linked stack overflow post was written by me 3 years ago :smile:

**Comment by rionda** [comment]
By the way, given how the algorithm works, one could temporarily store the scores as `int`s in a different vector, and only divide them by the sample size `r` and copy them to `scoreData` after the algorithm is done sampling all the shortest path, possibly in a `G.parallelForNodes` construct. The advantage would be that you can use `fetch_add` on an integral type now
